vm/systemSmartContracts/esdt.go

```diff
@@ -1326,82 +1440,110 @@ func (e *esdt) checkSpecialRolesAccordingToTokenType(args [][]byte, token *ESDTD
 	return nil
 }
 
-func (e *esdt) setSpecialRole(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	returnCode := e.checkArgumentsForSpecialRoleChanges(args)
-	if returnCode != vmcommon.Ok {
-		return returnCode
-	}
-
-	token, returnCode := e.basicOwnershipChecks(args)
-	if returnCode != vmcommon.Ok {
-		return returnCode
-	}
+func (e *esdt) setRolesForTokenAndAddress(
+	token *ESDTDataV2,
+	address []byte,
+	roles [][]byte,
+) (bool, bool, vmcommon.ReturnCode) {
```

Comment:
this is prone to programming errors. Strong suggestion to use a simple, un-exported DTO

Reply:
done